content/blog/2019-10-08-remote-read-meets-streaming.md

```diff
@@ -0,0 +1,301 @@
+message QueryResult {
+  repeated prometheus.TimeSeries timeseries = 1;
+}
+```
+
+This essentially allows read access to metrics in TSDB that Prometheus collected. The main use cases for remote read are:
+
+* Seamless Prometheus upgrades between different data formats of Prometheus, so having [Prometheus reading from another Prometheus](https://www.robustperception.io/accessing-data-from-prometheus-1-x-in-prometheus-2-0) 
+* Prometheus being able to read from 3rd party long term storage systems e.g InfluxDB.
+* 3rd party system querying data from Prometheus e.g [Thanos](https://thanos.io).
+
+## Remote Read: Problem Statement
+
+There were two key problems for such a simple remote read. It was easy to use and understand, but there were no
+streaming capabilities within single HTTP request, for the protobuf format we defined. Secondly, the response was 
+including raw samples (`float64` value and `int64` timestamp) instead of
+an encoded, compressed batch of samples called "chunks" that are used to store metrics inside TSDB.
+
+The server algorithm for remote read before improvement was follows: 
+
+1. Parse request.
+1. Select metrics from TSDB.
+1. For all decoded series:
+  * For all samples:
+      * Add to response protobuf
+1. Marshal response.
+1. Snappy compress.
+1. Send back the HTTP response.  
+ 
+In the essence, this means that the whole response of the remote read had to be buffered in raw, uncompressed form to marshall it 
+in potentially huge protobuf before sending it to the client. The whole response has to then be fully buffered in the client again to be able
+to unmarshal it from protobuf. Only after that client could access raw samples.
+
+What does it mean? It means that requests for, let's say, only 8 hours that matches 10 000 series can take up to **2.5GB** of memory allocated by both client and server each!
+
+Find below, memory usage metric for both Prometheus and [Thanos Sidecar](https://thanos.io/components/sidecar.md/) (remote read client) during remote read request time:
+
+![Prometheus 2.12.0: RSS of single read 8h of 10k series](/assets/blog/2019-10-08/10kseries8hours-2.12.png)
+
+![Prometheus 2.12.0: Heap-only allocations of single read 8h of 10k series](/assets/blog/2019-10-08/10series8hours-2.12-allocs.png)
+
+It's worth to note, that querying 10k series is never great even for Prometheus native HTTP `query_range` endpoint,
+simply because fetching and storing hundreds of megabytes in your browser is not great. 
+Additionally, for dashboards and rendering purposes it is not practical to have that much of the data. 
+That's why usually we craft queries that have no more than 20 series.
+
+This is great, but a very common technique is to compose queries in such way that query returns **aggregated** 20 series, 
+however underneath the query engine has to touch potentially thousands of series to evaluate the response (e.g `count()` function or [subqueries](https://prometheus.io/blog/2019/01/28/subquery-support/)).
```

Comment:
I'm not sure count and subqueries are good examples here. Neither inherently touch lots of series.

Reply:
Why not? They are in form of map reduce that can give you single series but touching potentially millions.

Any good alternative then? (:

Reply:
Subqueries are across time, not across series. Aggregators are a better example.